src/hibernate-resume/hibernate-resume-generator.c

```diff
@@ -81,6 +98,90 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
+                { "id",           JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, id),            JSON_PERMISSIVE|JSON_DEBUG },
+                { "imageId",      JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, image_id),      JSON_PERMISSIVE|JSON_DEBUG },
+                { "versionId",    JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, version_id),    JSON_PERMISSIVE|JSON_DEBUG },
+                { "imageVersion", JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, image_version), JSON_PERMISSIVE|JSON_DEBUG },
+                {},
+        };
+
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_free_ char *location_str = NULL, *device = NULL, *id = NULL, *image_id = NULL,
+                       *version_id = NULL, *image_version = NULL;
+        char node[STRLEN("UUID=") + SD_ID128_UUID_STRING_MAX];
+        EFIHibernateLocation location = {};
+        int log_level, log_level_ignore;
+
+        log_level = arg_resume_device ? LOG_DEBUG : LOG_ERR;
+        log_level_ignore = arg_resume_device ? LOG_DEBUG : LOG_WARNING;
+
+        r = efi_get_variable_string(EFI_SYSTEMD_VARIABLE(HibernateLocation), &location_str);
+        if (r == -ENOENT) {
+                log_debug("EFI variable HibernateLocation is not set, skipping.");
+                return 0;
+        }
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to get EFI variable HibernateLocation: %m");
+
+        r = json_parse(location_str, 0, &v, NULL, NULL);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to parse HibernateLocation JSON object: %m");
+
+        r = json_dispatch(v, dispatch_table, NULL, JSON_LOG, &location);
+        if (r < 0)
+                return r;
+
+        xsprintf(node, "UUID=" SD_ID128_UUID_FORMAT_STR, SD_ID128_FORMAT_VAL(location.uuid));
+
+        device = fstab_node_to_udev_node(node);
+        if (!device)
+                return log_full_errno(log_level, SYNTHETIC_ERRNO(ENODEV), "Failed to get resume device through HibernateLocation.");
+
+        r = parse_os_release(NULL,
+                             "ID", &id,
+                             "IMAGE_ID", &image_id,
+                             "VERSION_ID", &version_id,
+                             "IMAGE_VERSION", &image_version);
+        if (r < 0)
+                log_full_errno(log_level_ignore, r, "Failed to parse os-release, ignoring: %m");
+
+        if (!streq_ptr(id, location.id) ||
+            !streq_ptr(image_id, location.image_id) ||
+            !streq_ptr(version_id, location.version_id) ||
+            !streq_ptr(image_version, location.image_version)) {
+
+                log_notice("HibernateLocation system info mismatches with os-release, not resuming from it.");
+                return 0;
+        }
```

Comment:
please do this check before you start formatting the device, since conceptually it jus tmake sense to do the check first, and then use the data